spec/support/jazz.rb

```diff
@@ -0,0 +1,193 @@
+      end
+    end
+
+    def self.to_id(object)
+      "#{object.class.name.split("::").last}/#{object.name}"
+    end
+
+    def self.find(id)
+      class_name, object_name = id.split("/")
+      Models.data[class_name].find { |obj| obj.name == object_name }
+    end
+  end
+
+  # Here's a new-style GraphQL type definition
+  class Ensemble < BaseObject
+    implements GloballyIdentifiable
+    description "A group of musicians playing together"
+    config :config, :configged
+    field :name, "String", null: false
+    field :musicians, "[Jazz::Musician]", null: false
+    # Test extra arguments:
+    field :upcaseName, String, null: false, upcase: true
+
+    def upcase_name
+      @object.name # upcase is applied by the superclass
+    end
+  end
+
+  class Family < GraphQL::Enum
+    description "Groups of musical instruments"
+    # support string and symbol
+    value "STRING", "Makes a sound by vibrating strings", value: :str
+    value :WOODWIND, "Makes a sound by vibrating air in a pipe"
+    value :BRASS, "Makes a sound by amplifying the sound of buzzing lips"
+    value "PERCUSSION", "Makes a sound by hitting something that vibrates"
+    value "KEYS", "Neither here nor there, really"
+    value "DIDGERIDOO", "Makes a sound by amplifying the sound of buzzing lips", deprecation_reason: "Merged into BRASS"
+  end
+
+  # Lives side-by-side with an old-style definition
+  using GraphQL::DeprecatedDSL # for ! and types[]
+  InstrumentType = GraphQL::ObjectType.define do
+    name "Instrument"
+    implements GloballyIdentifiable
+    field :id, !types.ID, "A unique identifier for this object", resolve: ->(obj, args, ctx) { GloballyIdentifiable.to_id(obj) }
+    field :name, !types.String
+    if RUBY_ENGINE == "jruby"
+      # JRuby doesn't support refinements, so the `using` above won't work
+      field :family, Family.to_non_null_type
+    else
+      field :family, !Family
+    end
+  end
+
+  class Musician < BaseObject
+    implements GloballyIdentifiable
+    description "Someone who plays an instrument"
+    field :name, String, null: false
+    field :instrument, InstrumentType, null: false
+  end
+
+  # Another new-style definition, with method overrides
+  class Query < BaseObject
+    field :ensembles, [Ensemble], null: false
+    field :find, GloballyIdentifiable, null: true do
+      argument :id, "ID", null: false
+    end
+    field :instruments, [InstrumentType], null: false do
+      argument :family, Family, null: true
+    end
+
+    def ensembles
+      Models.data["Ensemble"]
+    end
+
+    def find(id:)
+      GloballyIdentifiable.find(id)
+    end
+
+    def instruments(family: nil)
+      objs = Models.data["Instrument"]
+      if family
+        objs = objs.select { |i| i.family == family }
+      end
+      objs
+    end
+  end
+
+  class EnsembleInput < GraphQL::InputObject
+    argument :name, String, null: false
+  end
+
+  class Mutation < BaseObject
+    field :addEnsemble, Ensemble, null: false do
```

Comment:
Would it make sense to support `field :add_ensemble, ...` which would automatically get camel-cased?

Reply:
I was torn about _where_ the renaming should happen. For example, this field is called `:addEnsemble` here, but it calls the `#add_ensemble` method below.

One thing I want to take care of is compatibility: some people _already_ have underscored APIs, and I want them to be able to migrate to this API.

Can we accomplish that somehow, providing a user-friendly auto-camelize feature, while supporting underscore names as needed? Here are some thoughts: 

- Treat strings and symbols differently: symbols are camelized, but if a string is given, assume it has already been prepared, so don't auto-camelize it. 
- Make the auto-camelize function configurable, so it can be disabled 

Anything else?

Reply:
I’m not a huge fan of the string vs symbol behavior, but I could get behind it anyway.

But, could camelization be handled through some sort of instrumentation? If you could change the field’s name that way, without it changing any of the resolution behavior (ie, it would still try to resolve against the underscore name), that seems like it’d be ideal.

Or the library could camelize by default, but you can specify `camelize: false` on the field to bypass.

I do think it’s important that camelizing field names be something that is at _least_ opt-out, and maybe the safest is to make it opt-in.

Reply:
We might be able to [follow along the lines of Graphene](http://docs.graphene-python.org/en/latest/types/schema/#auto-camelcase-field-names):

> By default all field and argument names (that are not explicitly set with the `name` arg) will be converted from `snake_case` to `camelCase`
> 
> [...]
> 
> To disable this behavior, set the `auto_camelcase` to `False` upon schema instantiation.
> ```python
> my_schema = Schema(
>     query=MyRootQuery,
>     auto_camelcase=False,
> )
> ```

Reply:
🤔  I'll give it a try ...